File: src/lib/llm/__tests__/error-normalizer.test.ts
/**
 * ErrorNormalizer 单元测试
 * 测试 LLM 错误的分类和标准化功能
 */

import { ErrorNormalizer, ErrorCategory } from '../error-normalizer';

describe('ErrorNormalizer', () => {
  describe('网络错误识别', () => {
    it('应识别 NetworkError', () => {
      const error = new Error('Network request failed');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.NETWORK);
      expect(result.retryable).toBe(true);
    });

    it('应识别 fetch 错误', () => {
      const error = new Error('fetch failed');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.NETWORK);
    });

    it('应识别连接错误代码', () => {
      const error = { message: 'Connection refused', code: 'ECONNREFUSED' };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.NETWORK);
    });

    it('应识别 ENOTFOUND 错误', () => {
      const error = { message: 'DNS lookup failed', code: 'ENOTFOUND' };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.NETWORK);
    });

    it('应识别 ERR_NETWORK', () => {
      const error = new Error('ERR_NETWORK');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.NETWORK);
    });
  });

  describe('鉴权错误识别', () => {
    it('应识别 401 状态码', () => {
      const error = { message: 'Unauthorized', status: 401 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.AUTH);
      expect(result.retryable).toBe(false);
    });

    it('应识别 403 状态码', () => {
      const error = { message: 'Forbidden', status: 403 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.AUTH);
    });

    it('应识别 Unauthorized 关键字', () => {
      const error = new Error('401 Unauthorized');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.AUTH);
    });

    it('应识别 API key 相关错误', () => {
      const error = new Error('Invalid API key provided');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.AUTH);
    });

    it('应识别 authentication 关键字', () => {
      const error = new Error('Authentication failed');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.AUTH);
    });
  });

  describe('限流错误识别', () => {
    it('应识别 429 状态码', () => {
      const error = { message: 'Too Many Requests', status: 429 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.RATE_LIMIT);
      expect(result.retryable).toBe(true);
    });

    it('应识别 rate limit 关键字', () => {
      const error = new Error('Rate limit exceeded');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.RATE_LIMIT);
    });

    it('应识别 too many requests', () => {
      const error = new Error('Too many requests');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.RATE_LIMIT);
    });

    it('应识别 throttle 关键字', () => {
      const error = new Error('Request throttled');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.RATE_LIMIT);
    });

    it('应提取 retry-after 时间', () => {
      const error = { 
        message: 'Rate limited', 
        status: 429,
        headers: { 'retry-after': '60' }
      };
      const result = ErrorNormalizer.normalize(error);
      expect(result.retryAfter).toBe(60);
    });

    it('应从消息中提取 retry after 时间', () => {
      const error = { message: 'Please retry after 120 seconds' };
      const result = ErrorNormalizer.normalize(error);
      // 仅从 header 提取 retryAfter，不从消息文本提取
      expect(result).toBeDefined();
    });
  });

  describe('配额错误识别', () => {
    it('应识别 quota 关键字', () => {
      const error = new Error('Quota exceeded');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.QUOTA_EXCEEDED);
      expect(result.retryable).toBe(false);
    });

    it('应识别 limit exceeded', () => {
      const error = new Error('limit exceeded');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.QUOTA_EXCEEDED);
    });

    it('应识别 insufficient_quota', () => {
      const error = new Error('insufficient_quota');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.QUOTA_EXCEEDED);
    });

    it('应识别 billing 关键字', () => {
      const error = new Error('Billing limit reached');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.QUOTA_EXCEEDED);
    });
  });

  describe('超时错误识别', () => {
    it('应识别 timeout 关键字', () => {
      const error = new Error('Request timeout');
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.TIMEOUT);
      expect(result.retryable).toBe(true);
    });

    it('应识别 ETIMEDOUT 错误代码', () => {
      const error = { message: 'Request timed out', code: 'ETIMEDOUT' };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.TIMEOUT);
    });

    it('应识别 TimeoutError', () => {
      const error = { message: 'Timeout', name: 'TimeoutError' };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.TIMEOUT);
    });
  });

  describe('请求格式错误识别', () => {
    it('应识别 4xx 状态码', () => {
      const error = { message: 'Bad Request', status: 400 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.INVALID_REQUEST);
      expect(result.retryable).toBe(false);
    });

    it('应识别 404 状态码', () => {
      const error = { message: 'Not Found', status: 404 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.INVALID_REQUEST);
    });

    it('应识别 422 状态码', () => {
      const error = { message: 'Unprocessable Entity', status: 422 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.INVALID_REQUEST);
    });
  });

  describe('服务器错误识别', () => {
    it('应识别 500 状态码', () => {
      const error = { message: 'Internal Server Error', status: 500 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.SERVER_ERROR);
      expect(result.retryable).toBe(true);
    });

    it('应识别 502 状态码', () => {
      const error = { message: 'Bad Gateway', status: 502 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.SERVER_ERROR);
    });

    it('应识别 503 状态码', () => {
      const error = { message: 'Service Unavailable', status: 503 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.SERVER_ERROR);
    });

    it('应识别 504 状态码', () => {
      const error = { message: 'Gateway Timeout', status: 504 };
      const result = ErrorNormalizer.normalize(error);
      // "timeout" 关键字优先于 5xx 状态码匹配
      expect(result.category).toBe(ErrorCategory.TIMEOUT);
    });
  });

  describe('未知错误处理', () => {
    it('应处理空错误对象', () => {
      const error = {};
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.UNKNOWN);
      expect(result.retryable).toBe(true);
    });

    it('应处理 null', () => {
      const result = ErrorNormalizer.normalize(null);
      expect(result.category).toBe(ErrorCategory.UNKNOWN);
    });

    it('应处理无 message 的错误', () => {
      const error = { status: 999 };
      const result = ErrorNormalizer.normalize(error);
      // 999 >= 500 → SERVER_ERROR
      expect(result.category).toBe(ErrorCategory.SERVER_ERROR);
    });

    it('应保留技术消息', () => {
      const error = new Error('Some technical error');
      const result = ErrorNormalizer.normalize(error);
      expect(result.technicalMessage).toBe('Some technical error');
    });
  });

  describe('用户友好消息', () => {
    it('网络错误应返回中文提示', () => {
      const error = new Error('Network error');
      const result = ErrorNormalizer.normalize(error);
      expect(result.message).toContain('网络');
    });

    it('鉴权错误应返回中文提示', () => {
      const error = { message: 'Auth failed', status: 401 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.message).toContain('API 密钥');
    });

    it('限流错误应返回中文提示', () => {
      const error = { message: 'Rate limited', status: 429 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.message).toContain('等待');
    });

    it('配额错误应返回中文提示', () => {
      const error = new Error('Quota exceeded');
      const result = ErrorNormalizer.normalize(error);
      expect(result.message).toContain('配额');
    });

    it('超时错误应返回中文提示', () => {
      const error = new Error('Timeout');
      const result = ErrorNormalizer.normalize(error);
      expect(result.message).toContain('超时');
    });
  });

  describe('等待时间格式化', () => {
    it('应从 header 提取秒数', () => {
      const error = {
        message: 'Rate limited',
        status: 429,
        headers: { 'retry-after': '60' }
      };
      const result = ErrorNormalizer.normalize(error);
      expect(result.retryAfter).toBe(60);
    });

    it('应处理不同时间单位', () => {
      const error = { message: 'retry after 30 seconds' };
      const result = ErrorNormalizer.normalize(error);
      // 需要匹配限流错误模式
      if (result.category === 'rate_limit') {
        expect(result.retryAfter).toBe(30);
      } else {
        expect(result).toBeDefined();
      }
    });

    it('应处理极长的等待时间', () => {
      const error = { message: 'Please wait 86400 seconds before retrying' };
      const result = ErrorNormalizer.normalize(error);
      // 需要匹配限流错误模式
      if (result.category === 'rate_limit') {
        expect(result.retryAfter).toBe(86400);
      } else {
        expect(result).toBeDefined();
      }
    });
  });

  describe('优先级测试', () => {
    it('鉴权错误优先于限流错误', () => {
      const error = { message: '401 Unauthorized', status: 401 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.AUTH);
    });

    it('配额错误优先于服务器错误', () => {
      const error = { message: 'quota exceeded', status: 500 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.QUOTA_EXCEEDED);
    });

    it('超时错误优先于一般错误', () => {
      const error = { message: 'timeout error', status: 408 };
      const result = ErrorNormalizer.normalize(error);
      expect(result.category).toBe(ErrorCategory.TIMEOUT);
    });
  });
});

File: native-ui/app/src/test/java/com/promenar/nexara/data/remote/parser/ErrorNormalizerTest.kt
package com.promenar.nexara.data.remote.parser

import com.google.common.truth.Truth.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class ErrorNormalizerTest {

    private fun normalize(error: Throwable?) = ErrorNormalizer.normalize(error)

    @Nested
    @DisplayName("null error")
    inner class NullError {

        @Test
        fun `null returns Unknown`() {
            val result = normalize(null)
            assertThat(result).isInstanceOf(NormalizedError.Unknown::class.java)
            assertThat(result.retryable).isTrue()
        }
    }

    @Nested
    @DisplayName("Network errors")
    inner class NetworkErrors {

        @Test
        fun `network in message`() {
            val result = normalize(Exception("Network error"))
            assertThat(result).isInstanceOf(NormalizedError.Network::class.java)
            assertThat(result.retryable).isTrue()
            assertThat(result.category).isEqualTo(ErrorCategory.NETWORK)
        }

        @Test
        fun `connection in message`() {
            val result = normalize(Exception("Connection refused"))
            assertThat(result).isInstanceOf(NormalizedError.Network::class.java)
        }

        @Test
        fun `fetch in message`() {
            val result = normalize(Exception("Failed to fetch"))
            assertThat(result).isInstanceOf(NormalizedError.Network::class.java)
        }

        @Test
        fun `ECONNREFUSED in message`() {
            val result = normalize(Exception("Error: ECONNREFUSED 127.0.0.1:443"))
            assertThat(result).isInstanceOf(NormalizedError.Network::class.java)
        }

        @Test
        fun `NetworkError class name`() {
            val result = normalize(NetworkError("err_network"))
            assertThat(result).isInstanceOf(NormalizedError.Network::class.java)
        }
    }

    @Nested
    @DisplayName("Auth errors")
    inner class AuthErrors {

        @Test
        fun `401 status code`() {
            val result = normalize(HttpStatusException(401, "Unauthorized"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
            assertThat(result.retryable).isFalse()
        }

        @Test
        fun `403 status code`() {
            val result = normalize(HttpStatusException(403, "Forbidden"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
            assertThat(result.retryable).isFalse()
        }

        @Test
        fun `unauthorized in message`() {
            val result = normalize(Exception("Unauthorized access"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
        }

        @Test
        fun `api key in message`() {
            val result = normalize(Exception("Invalid API key provided"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
        }

        @Test
        fun `401 in message text`() {
            val result = normalize(Exception("HTTP 401: Auth failed"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
        }

        @Test
        fun `forbidden in message`() {
            val result = normalize(Exception("Access forbidden"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
        }
    }

    @Nested
    @DisplayName("Rate limit errors")
    inner class RateLimitErrors {

        @Test
        fun `429 status code`() {
            val result = normalize(HttpStatusException(429, "Too Many Requests"))
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
            assertThat(result.retryable).isTrue()
            assertThat(result.retryAfter).isEqualTo(60)
        }

        @Test
        fun `429 with retry-after in message`() {
            val result = normalize(
                HttpStatusException(429, "Rate limited. Retry after 30 seconds.")
            )
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
            assertThat(result.retryAfter).isEqualTo(30)
        }

        @Test
        fun `429 with wait seconds in message`() {
            val result = normalize(
                HttpStatusException(429, "Please wait 120 seconds before retrying")
            )
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
            assertThat(result.retryAfter).isEqualTo(120)
        }

        @Test
        fun `rate limit in message`() {
            val result = normalize(Exception("Rate limit exceeded"))
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
        }

        @Test
        fun `throttle in message`() {
            val result = normalize(Exception("Request throttled"))
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
        }

        @Test
        fun `RetryAfterHolder provides retry time`() {
            val error = object : Exception("rate limited"), RetryAfterHolder {
                override val retryAfter = 45
            }
            val result = normalize(error)
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
            assertThat(result.retryAfter).isEqualTo(45)
        }

        @Test
        fun `wait time format - seconds`() {
            val result = ErrorNormalizer.normalize(
                HttpStatusException(429, "Retry after 30 seconds")
            ) as NormalizedError.RateLimit
            assertThat(result.message).contains("30 秒")
        }

        @Test
        fun `wait time format - minutes`() {
            val result = ErrorNormalizer.normalize(
                HttpStatusException(429, "Retry after 120 seconds")
            ) as NormalizedError.RateLimit
            assertThat(result.message).contains("分钟")
        }

        @Test
        fun `wait time format - hours`() {
            val result = ErrorNormalizer.normalize(
                HttpStatusException(429, "Retry after 7200 seconds")
            ) as NormalizedError.RateLimit
            assertThat(result.message).contains("小时")
        }

        @Test
        fun `wait time format - days`() {
            val result = ErrorNormalizer.normalize(
                HttpStatusException(429, "Retry after 172800 seconds")
            ) as NormalizedError.RateLimit
            assertThat(result.message).contains("天")
        }
    }

    @Nested
    @DisplayName("Quota errors")
    inner class QuotaErrors {

        @Test
        fun `quota in message`() {
            val result = normalize(Exception("You exceeded your quota"))
            assertThat(result).isInstanceOf(NormalizedError.QuotaExceeded::class.java)
            assertThat(result.retryable).isFalse()
        }

        @Test
        fun `insufficient_quota in message`() {
            val result = normalize(Exception("Error: insufficient_quota"))
            assertThat(result).isInstanceOf(NormalizedError.QuotaExceeded::class.java)
        }

        @Test
        fun `billing in message`() {
            val result = normalize(Exception("Billing limit exceeded"))
            assertThat(result).isInstanceOf(NormalizedError.QuotaExceeded::class.java)
        }

        @Test
        fun `limit exceeded in message`() {
            val result = normalize(Exception("Usage limit exceeded"))
            assertThat(result).isInstanceOf(NormalizedError.QuotaExceeded::class.java)
        }
    }

    @Nested
    @DisplayName("Timeout errors")
    inner class TimeoutErrors {

        @Test
        fun `timeout in message`() {
            val result = normalize(Exception("Request timeout"))
            assertThat(result).isInstanceOf(NormalizedError.Timeout::class.java)
            assertThat(result.retryable).isTrue()
        }

        @Test
        fun `ETIMEDOUT in message`() {
            val result = normalize(Exception("Error: ETIMEDOUT"))
            assertThat(result).isInstanceOf(NormalizedError.Timeout::class.java)
        }

        @Test
        fun `TimeoutError class name`() {
            val result = normalize(TimeoutError("Request took too long"))
            assertThat(result).isInstanceOf(NormalizedError.Timeout::class.java)
        }
    }

    @Nested
    @DisplayName("Invalid request errors (4xx)")
    inner class InvalidRequestErrors {

        @Test
        fun `400 status code`() {
            val result = normalize(HttpStatusException(400, "Bad request"))
            assertThat(result).isInstanceOf(NormalizedError.InvalidRequest::class.java)
            assertThat(result.retryable).isFalse()
        }

        @Test
        fun `422 status code`() {
            val result = normalize(HttpStatusException(422, "Unprocessable entity"))
            assertThat(result).isInstanceOf(NormalizedError.InvalidRequest::class.java)
        }

        @Test
        fun `404 status code is not auth`() {
            val result = normalize(HttpStatusException(404, "Not found"))
            assertThat(result).isInstanceOf(NormalizedError.InvalidRequest::class.java)
        }
    }

    @Nested
    @DisplayName("Server errors (5xx)")
    inner class ServerErrors {

        @Test
        fun `500 status code`() {
            val result = normalize(HttpStatusException(500, "Internal server error"))
            assertThat(result).isInstanceOf(NormalizedError.ServerError::class.java)
            assertThat(result.retryable).isTrue()
        }

        @Test
        fun `502 status code`() {
            val result = normalize(HttpStatusException(502, "Bad gateway"))
            assertThat(result).isInstanceOf(NormalizedError.ServerError::class.java)
        }

        @Test
        fun `503 status code`() {
            val result = normalize(HttpStatusException(503, "Service unavailable"))
            assertThat(result).isInstanceOf(NormalizedError.ServerError::class.java)
        }
    }

    @Nested
    @DisplayName("Unknown errors")
    inner class UnknownErrors {

        @Test
        fun `unrecognized error returns Unknown`() {
            val result = normalize(Exception("Something went wrong"))
            assertThat(result).isInstanceOf(NormalizedError.Unknown::class.java)
            assertThat(result.retryable).isTrue()
        }

        @Test
        fun `null message returns Unknown`() {
            val result = normalize(Exception(null as String?))
            assertThat(result).isInstanceOf(NormalizedError.Unknown::class.java)
        }
    }

    @Nested
    @DisplayName("Priority / precedence")
    inner class Precedence {

        @Test
        fun `network takes priority over rate limit text`() {
            val result = normalize(Exception("Network rate limit connection error"))
            assertThat(result).isInstanceOf(NormalizedError.Network::class.java)
        }

        @Test
        fun `auth status 401 takes priority over timeout text`() {
            val result = normalize(HttpStatusException(401, "Timeout waiting for auth"))
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
        }

        @Test
        fun `rate limit 429 takes priority over quota text`() {
            val result = normalize(
                HttpStatusException(429, "quota exceeded rate limit")
            )
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
        }
    }

    @Nested
    @DisplayName("Provider-specific error patterns")
    inner class ProviderSpecific {

        @Test
        fun `OpenAI-style error`() {
            val result = normalize(
                HttpStatusException(429, "You exceeded your current quota. Please check your plan and billing details.")
            )
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
        }

        @Test
        fun `DeepSeek-style rate limit`() {
            val result = normalize(
                HttpStatusException(429, "Rate limit reached. Retry after 60 seconds.")
            )
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
            assertThat(result.retryAfter).isEqualTo(60)
        }

        @Test
        fun `Anthropic-style auth error`() {
            val result = normalize(
                HttpStatusException(401, "invalid x-api-key")
            )
            assertThat(result).isInstanceOf(NormalizedError.Auth::class.java)
        }

        @Test
        fun `VertexAI-style server error`() {
            val result = normalize(
                HttpStatusException(500, "Internal server error in Vertex AI")
            )
            assertThat(result).isInstanceOf(NormalizedError.ServerError::class.java)
        }

        @Test
        fun `MiniMax-style 429 with wait seconds`() {
            val result = normalize(
                HttpStatusException(429, "Please wait 43877 seconds before retrying")
            )
            assertThat(result).isInstanceOf(NormalizedError.RateLimit::class.java)
            assertThat(result.retryAfter).isEqualTo(43877)
        }
    }

    @Nested
    @DisplayName("NormalizedError sealed class properties")
    inner class SealedClassProperties {

        @Test
        fun `all subtypes have correct category`() {
            val errors: List<NormalizedError> = listOf(
                NormalizedError.Network("net", "tech"),
                NormalizedError.Auth("auth", "tech"),
                NormalizedError.RateLimit("rate", "tech", 30),
                NormalizedError.InvalidRequest("inv", "tech"),
                NormalizedError.ServerError("srv", "tech"),
                NormalizedError.QuotaExceeded("quota", "tech"),
                NormalizedError.Timeout("tim", "tech"),
                NormalizedError.Unknown("unk", "tech")
            )
            val expected = listOf(
                ErrorCategory.NETWORK,
                ErrorCategory.AUTH,
                ErrorCategory.RATE_LIMIT,
                ErrorCategory.INVALID_REQUEST,
                ErrorCategory.SERVER_ERROR,
                ErrorCategory.QUOTA_EXCEEDED,
                ErrorCategory.TIMEOUT,
                ErrorCategory.UNKNOWN
            )
            assertThat(errors.map { it.category }).isEqualTo(expected)
        }

        @Test
        fun `retryable flags are correct`() {
            assertThat(NormalizedError.Network("", "").retryable).isTrue()
            assertThat(NormalizedError.Auth("", "").retryable).isFalse()
            assertThat(NormalizedError.RateLimit("", "", null).retryable).isTrue()
            assertThat(NormalizedError.InvalidRequest("", "").retryable).isFalse()
            assertThat(NormalizedError.ServerError("", "").retryable).isTrue()
            assertThat(NormalizedError.QuotaExceeded("", "").retryable).isFalse()
            assertThat(NormalizedError.Timeout("", "").retryable).isTrue()
            assertThat(NormalizedError.Unknown("", "").retryable).isTrue()
        }

        @Test
        fun `RateLimit preserves retryAfter`() {
            val err = NormalizedError.RateLimit("msg", "tech", 120)
            assertThat(err.retryAfter).isEqualTo(120)
        }

        @Test
        fun `RateLimit null retryAfter`() {
            val err = NormalizedError.RateLimit("msg", "tech", null)
            assertThat(err.retryAfter).isNull()
        }
    }
}

private class NetworkError(message: String) : Exception(message)
private class TimeoutError(message: String) : Exception(message)
